NCCL Profiler Plugin API updates proposal

spease-fb opened the pull request:
Differential Revision: D41634378

facebook-github-bot committed:
Re-sync with internal repository (#3)

Co-authored-by: Facebook Community Bot <6422482+facebook-github-bot@users.noreply.github.com>

Commit:
Set up ShipIt

fbshipit-source-id: 6939f9f252d3048c174c7cacabcc15f921b6096c

minsii committed:
move ntrace patch to third-party (exclude ext_tag) (#4)

Summary:
Pull Request resolved: https://github.com/facebookresearch/nccl/pull/4

Introduce ntrace into NCCL to trace important IB events per collective communication. It requires a separate module ntrace_rt to enable in Makefile based build. For now, ENABLE_NTRACE is always off when building from Makefile.

Co-author: Xianghuai Zhang <xianghuaizhang@fb.com>

Reviewed By: kingchc

Differential Revision: D38637718

fbshipit-source-id: aca15a9297f175aac873dbfffef3845fcd596b87

Commit:
Make ntrace_rt identifers more compliant with Meta conventions

Summary:
Refactoring done by running:
```
arc cast //tools/cAST:fbtidy -paths ../third-party/ntrace_rt --header-filter=".*" -keep-going
```

and then manually fixing things it screwed up on.

Reviewed By: minsii

Differential Revision: D40281083

fbshipit-source-id: d080dfe45ca843fb13ba00c720015f742d8de981

sjeaugey committed:
Upgrade exp (#9)

Summary: Pull Request resolved: https://github.com/facebookresearch/nccl/pull/9

Reviewed By: wesbland

Differential Revision: D40610635

Pulled By: kingchc

fbshipit-source-id: c06551709d1a59ccc89088a97366f6f94cb16224

minsii commented on src/proxy.cc:
```diff
@@ -580,7 +580,7 @@ static ncclResult_t ncclProxyGetPostedOps(struct ncclComm* comm, int* added) {
       NCCLCHECK(ncclProfEventStart(comm->proxyState.profilerRecord, &profId, PROF_TYPE_SLEEP, 0));
       pthread_cond_wait(&pool->cond, &pool->mutex);
-      ncclProfEventTime(comm->proxyState.profilerRecord, profId, NCCL_PROF_TS_END);
+      NCCLCHECK(ncclProfEventTime(comm->proxyState.profilerRecord, profId, NCCL_PROF_TS_END));
```
Is the NCCLCHECK change a bug fix which should be moved to the other commit?

spease-fb replied:
No - when I converted ncclProfEventTime, I also changed it to return a success error code.

Design justification-

Since the return comes after the timestamp is logged and this should only add one branch to check the result, I figured any change to results should be nonexistent to negligible, but this allows for more flexibility in the plugin if it needs it. It's also consistent with the other API functions which have more potential to fail (ncclProfEventTime seems to be solely for recording a single timestamp, so it's probably the most deterministic part of the new profiler API).

I did consider deliberately leaving ncclProfEventTime as a built-in function so that you would have ncclProfEventStart (customizable but plugin call overhead) and ncclProfEventTime (not customizable but no overhead), but I believe this would implicitly force ncclProfEventStart to fill out the built-in data structures whether it's using them or not in order to prevent ncclProfEventTime from crashing.

I think we can use the default NVIDIA data structures, but I think the flexibility would be more important to us than sub-microsecond accuracy. Since this is a part of the plugin API that will be consumed by multiple people and will have to be backwards compatible, I decided in favor of the option that gave us the most flexibility rather than leaving it alone until absolutely necessary (at which point it might be more difficult to request a change because it would require a new version or breaking things for other people).

minsii commented on src/include/ibvwrap.h:
```diff
@@ -13,9 +13,12 @@
 
+extern struct ncclProfRecord* g_currentProfilerRecord;
+
 // Dynamically handle dependencies on IB verbs
```
Can we reorder the "Replace ntrace hooks with new profiler" commit be the last one, and add commit description about the pre-required changes that were introduced in previous commits? It would be good to also have a summary of each commit's purpose in the PR top description.

spease-fb replied:
Yes - I have attempted to re-export with a bit more descriptive metatext, or will open a new PR as need be.

spease-fb committed:
Apply NVIDIA profiler patch

Differential Revision: D41483497

fbshipit-source-id: 1c35079ec07f4c3f0f081293dd0a1745159eab70

spease-fb committed:
Fix segfaults in plugin load code

Summary:
This fixes segfaults that could occur with the original patch provided to us under the following conditions:
(1) The environment variable to switch on the profiler does not exist
(2) The profiler shared object library does not contain the expected symbol for the function table

Differential Revision: D41990362

fbshipit-source-id: a02742e6b245ff23e91122b0dc1ccbcbdde370bc

spease-fb committed:
Convert internal infiniband hooks to use new profiler API

Summary: This replaces the proprietary hooks in the infiniband submodule with the general hook functions provided by NVIDIA's new profiler API.

Differential Revision: D41564279

fbshipit-source-id: 0d4582a8177e6a2ad91151a5d18b58ebc979f44f

Commit:
Convert remaining nccl profiler functions to plugin functions (#11)

Summary:
Convert the remaining nccl profiler functions to functions which are provided by the plugin shared object. This modifies the general plugin API as laid out by NVIDIA.

Pull Request resolved: https://github.com/facebookresearch/nccl/pull/11

Differential Revision: D41634378

fbshipit-source-id: 90758e2fbd1318598b4a0ccb5a3f0adb2d42b058
